tests/pkcs11_test.c

```diff
@@ -15,24 +17,167 @@
+    }
+
+    CK_SLOT_ID_PTR p_slot_list = (CK_SLOT_ID_PTR)malloc(ul_slot_count * sizeof(CK_SLOT_ID));
+    if (p_slot_list == NULL) {
+        FAIL("ERROR: Could not allocate memory.\n");
+    }
+
+    rv = pkcs11_lib->function_list->C_GetSlotList(CK_FALSE, p_slot_list, &ul_slot_count);
+    if (rv != CKR_OK) {
+        free(p_slot_list);
+        FAIL("ERROR: Could not get the slot list.\n");
+    }
+
+    size_t counter = 0;
+    for (CK_ULONG i = 0; i < ul_slot_count; i++) {
+        CK_TOKEN_INFO curr_token_info;
+
+        rv = pkcs11_lib->function_list->C_GetTokenInfo(p_slot_list[i], &curr_token_info);
+        if (rv != CKR_OK) {
+            free(p_slot_list);
+            FAIL("ERROR: Could not get info about the token in slot %lu.\n", p_slot_list[i]);
+        }
+
+        if (memcmp(curr_token_info.serialNumber, tokenInfo.serialNumber, sizeof(tokenInfo.serialNumber)) == 0 &&
+            memcmp(curr_token_info.label, tokenInfo.label, sizeof(tokenInfo.label)) == 0) {
+            slot_id = p_slot_list[i];
+            counter++;
+        }
+    }
+
+    free(p_slot_list);
+
+    if (counter == 1) return slot_id;
+    if (counter > 1) {
+        FAIL("ERROR: Found multiple matching slots/tokens.\n");
+    }
+    FAIL("ERROR: Could not find a slot/token using --serial, or --token\n");
+}
+
+static void s_pkcs11_clear_softhsm(struct aws_pkcs11_lib *pkcs11_lib) {
+    char cmd[120] = {'\0'};
+    const char* token_dir = aws_string_c_str(s_pkcs11_tester.token_dir);
+    if (token_dir[s_pkcs11_tester.token_dir->len - 1] == '/') {
+        sprintf(cmd, "rm -rf %s*", token_dir);
+    } else {
+        sprintf(cmd, "rm -rf %s/*", token_dir);
+    }
+    printf("Executing command: %s", cmd);
+    system(cmd);
+    next_free_slot_id = 0;
+    // Reload the library
+    pkcs11_lib->function_list->C_Finalize(NULL_PTR);
+    pkcs11_lib->function_list->C_Initialize(NULL_PTR);
+}
+
+static CK_OBJECT_HANDLE s_pkcs11_create_key(struct aws_pkcs11_lib *pkcs11_lib,
+                                            CK_SLOT_ID slot_id,
+                                            const char* const label,
+                                            const char* const id,
+                                            CK_SESSION_HANDLE session) {
+    CK_OBJECT_HANDLE priv_key;
+    CK_ATTRIBUTE privateKeyTemplate[] = {
+        { CKA_LABEL,    (void *)label,  (CK_ULONG)strlen(label) },
+        { CKA_ID,       (void *)id,     (CK_ULONG)strlen(id) }
+    };
+    CK_MECHANISM mech = { CKM_DES_KEY_GEN, NULL_PTR, 0};
+
+    CK_RV rv = pkcs11_lib->function_list->C_GenerateKey(session, &mech, privateKeyTemplate, 2, &priv_key);
+    if (rv != CKR_OK) {
+        FAIL("C_GenerateKey fails: PKCS#11 error: %s (0x%08lX)", s_ckr_str(rv), rv);
+    }
+    return priv_key;
+}
+
+static uint64_t s_pkcs11_softhsm_create_slot(struct aws_pkcs11_lib *pkcs11_lib,
+                                        const char *const token_name,
```

Comment:
super trivial, but in our code base we never really do const const

Reply:
ok